### DISCUSSION TITLE
Solvers

### DISCUSSION THREAD

**Discussion by SteveDiamond** [cvxpy/cvxpy#1271] [General]
Use this thread to discuss your experience with the different solvers CVXPY supports. Any particularly nice features or potential improvements?

**Comment by rileyjmurray** [comment]
I'll just make some notes about opportunities for improving the C-implementations for some numerical solvers:

- ECOS: add a ["Mehrotra-corrector" for the exponential cone](https://link.springer.com/article/10.1007/s10107-021-01631-4). This should substantially improve the ability of ECOS to solve exponential cone programs reliably. Right now ECOS often terminates with a message "combined line search fails" for exponential cone programs. (see [my comment on the ECOS repo](https://github.com/embotech/ecos/issues/177#issuecomment-753571910)). Fair warning: implementing this involves some *really* nasty third directional derivatives described in the appendix of the paper I linked. I think the folks developing Hypatia may have already implemented this, so you can check with them to find an open-source implementation.
- ECOS_BB: this solver has correctness issues in how it manages the branch and bound tree. There is at least one problem in the CVXPY test suite where ECOS_BB returns an incorrect solution. There are several problems mentioned in CVXPY GitHub issues where ECOS_BB returns incorrect results. It would help if someone took up a project to audit the ECOS_BB implementation and try to find where things might be going wrong.
- CVXOPT: at each iteration, the main cone programming algorithm takes a constant fraction step size to the boundary defined by conic constraints (see Section 7.2, step 5, of [the CVXOPT tech report](http://www.seas.ucla.edu/~vandenbe/publications/coneprog.pdf)). Ideally, CVXOPT would step within some properly defined "neighborhood" of the central path. The standard neighborhood is the one-sided infinity-norm neighborhood. With the current implementation, CVXOPT can fail due to numerical issues on some linear programs. I have verified that LP solvers that *do* use the one-sided infinity-norm neighborhood can solve the LP's that make CVXOPT fail.
- CVXOPT: the algorithm used here computes an "affine scaling direction", then defines a "centering parameter", then computes a "combined search direction" that it uses to update the decision variable. This process can slightly be improved by checking if a full step in the affine scaling direction is feasible with respect to cone constraints. If that happens, then updating the decision variables to take a full step in that direction will *immediately* arrive at the globally optimal solution of the cone program. If CVXOPT were to take advantage of this then it would occasionally terminate a few iterations earlier (useful for large problems or SDP's) and with solutions of very high precision.

I don't think this thread is a great place to have full conversations about the points I mentioned above (partly because they don't involve CVXPY code!) but I wanted to make a note of these points in one place. If someone is interested they can contact me through either of the two emails given on my website. Of course, these discussions could be had in the ECOS/CVXOPT repos or the CVXOPT Google Group.

Edit: Oh there are threaded replies! Excellent. Okay, we can have some limited discussion here then, if people are interested.

**Reply by SteveDiamond** [reply]
Another thing I'd add is having some sort of CI license for the commercial solvers. This is a major pain point in our current testing process. We have to run the Gurobi, MOSEK, etc tests locally, which is much less reliable.

**Reply by SteveDiamond** [reply]
I have since learned that Gurobi and Cplex actually could be tested in CI using the restricted licenses. That's something we should do.
